### PR TITLE
Adding audience field to device flow token request

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR

Usage from flytectl

```
flytectl get projects --logger.level=6  --admin.authType DeviceFlow 
```

This was verified by @shahwar9 in his flyte auth0 deployment . Thanks @shahwar9 for finding the issue for passing the audience in the first device auth request 

Earlier we were relying on default audience field on the auth server for it to work.

AllowedAudience flyteAdmin config will restrict what tokens it can accept .

Adding Audience to public client config . Followup change in admin to return this data from that config

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue

fixes https://github.com/flyteorg/flyte/issues/2862

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by pmahindrakar-oss** [comment]
cc : @shahwar9

File: clients/go/admin/deviceflow/token_orchestrator.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ import (
 )
 
 const (
+	audience       = "audience"
 	cliendID       = "client_id"
```
Comment (katrogan) [inline clients/go/admin/deviceflow/token_orchestrator.go:22]:
```suggestion
	clientID       = "client_id"
```

**Review by katrogan — DISMISSED** [review]

File: clients/go/admin/deviceflow/config.go [hunk 2]
```diff
@@ -7,4 +7,5 @@ type Config struct {
 	TokenRefreshGracePeriod config.Duration `json:"refreshTime" pflag:",grace period from the token expiry after which it would refresh the token."`
 	Timeout                 config.Duration `json:"timeout" pflag:",amount of time the device flow should complete or else it will be cancelled."`
 	PollInterval            config.Duration `json:"pollInterval" pflag:",amount of time the device flow would poll the token endpoint if auth server doesn't return a polling interval. Okta and google IDP do return an interval'"`
+	Audience                string          `json:"audience" pflag:",audience value to be passed when requesting access token.Required when auth server hasn't been configured with default audience"`
```
Comment (EngHabu) [inline clients/go/admin/deviceflow/config.go:10]: Should this come from the GetPublicClient server API?

Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/config.go]: Makes sense . This makes admin control it better instead of allowing  any  value being passed from flytectl

Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/config.go]: Done

**Review by katrogan — APPROVED** [review]

File: clients/go/admin/deviceflow/token_orchestrator.go [hunk 3]
```diff
@@ -150,7 +154,7 @@ func (t TokenOrchestrator) FetchTokenFromAuthFlow(ctx context.Context) (*oauth2.
 	if len(t.ClientConfig.Scopes) > 0 {
 		scopes = strings.Join(t.ClientConfig.Scopes, " ")
 	}
-	daReq := DeviceAuthorizationRequest{ClientID: t.ClientConfig.ClientID, Scope: scopes}
+	daReq := DeviceAuthorizationRequest{ClientID: t.ClientConfig.ClientID, Scope: scopes, Audience: t.ClientConfig.Audience}
```
Comment (EngHabu) [inline clients/go/admin/deviceflow/token_orchestrator.go:157]: Shouldn't this attempt to get the audience from the GetPublicClientInfo endpoint first?

Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/token_orchestrator.go:157]: Its coming from GetPublicClientInfo object.
https://github.com/flyteorg/flyteidl/pull/314/files#diff-e9c73c8c26d2564bad2d6d4543f229c55470bd18b8cbb900832ec4a7b2e1ad70R20-R42

Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/token_orchestrator.go:157]: cc : @EngHabu

**Review by eapolinario — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
